### PR TITLE
wifi: Add separate WQ for RX and TX_DONE processing

### PR DESCRIPTION
**Pull request by krish2718**

### CODE REVIEW

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/rx.c [hunk 1]
```diff
@@ -163,6 +163,42 @@ enum wifi_nrf_status wifi_nrf_fmac_rx_cmd_send(struct wifi_nrf_fmac_dev_ctx *fma
...
+		goto out;
+	}
+
+	status = wifi_nrf_fmac_rx_event_process(fmac_dev_ctx,
+						config);
+
+	if (status != WIFI_NRF_STATUS_SUCCESS) {
+		wifi_nrf_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+				      "%s: wifi_nrf_fmac_rx_event_process failed\n",
+				      __func__);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/rx.c:190]: config needs to be freed here?

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/event.c [hunk 2]
```diff
@@ -527,13 +528,73 @@ wifi_nrf_fmac_data_event_process(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx,
...
+					sizeof(struct nrf_wifi_tx_buff_done));
+		if (!config) {
+			wifi_nrf_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+					      "%s: Failed to allocate memory (TX)\n",
+					      __func__);
+			status = WIFI_NRF_STATUS_FAIL;
+			break;
+		}
+		wifi_nrf_osal_mem_cpy(fmac_dev_ctx->fpriv->opriv,
+					config,
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/event.c:578]: Indentation

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/event.c [hunk 3]
```diff
@@ -527,13 +528,73 @@ wifi_nrf_fmac_data_event_process(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx,
...
+			sizeof(struct nrf_wifi_rx_buff));
+		if (!config) {
+			wifi_nrf_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+					      "%s: Failed to allocate memory (RX)\n",
+					      __func__);
+			status = WIFI_NRF_STATUS_FAIL;
+			break;
+		}
+		wifi_nrf_osal_mem_cpy(fmac_dev_ctx->fpriv->opriv,
+					config,
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/event.c:543]: Indentation

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 4]
```diff
@@ -134,6 +159,13 @@ static enum wifi_nrf_status wifi_nrf_fmac_deinit_rx(struct wifi_nrf_fmac_dev_ctx
 
 	fpriv = fmac_dev_ctx->fpriv;
 
+#ifdef CONFIG_NRF700X_RX_WQ_ENABLED
+	wifi_nrf_osal_tasklet_free(fmac_dev_ctx->fpriv->opriv,
+				     fmac_dev_ctx->rx_tasklet);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:164]: Indentation

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c [hunk 5]
```diff
@@ -1311,41 +1253,69 @@ enum wifi_nrf_status tx_init(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx)
...
+#ifdef CONFIG_NRF700X_TX_DONE_WQ_ENABLED
+tx_done_tasklet_free:
+	wifi_nrf_osal_tasklet_free(fpriv->opriv,
+				   fmac_dev_ctx->tx_done_tasklet);
+wakeup_client_q_free:
+	wifi_nrf_utils_q_free(fpriv->opriv, fmac_dev_ctx->tx_config.wakeup_client_q);
+#endif /* CONFIG_NRF700X_TX_DONE_WQ_ENABLED */
+tx_spin_lock_free:
+	wifi_nrf_osal_spinlock_free(fmac_dev_ctx->fpriv->opriv,
+					fmac_dev_ctx->tx_config.tx_lock);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:1293]: Indentation

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c [hunk 6]
```diff
@@ -1311,41 +1253,69 @@ enum wifi_nrf_status tx_init(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx)
...
+tx_buff_map_free:
+	wifi_nrf_osal_mem_free(fmac_dev_ctx->fpriv->opriv,
+					fmac_dev_ctx->tx_config.buf_pool_bmp_p);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:1296]: Indentation

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c [hunk 7]
```diff
@@ -1311,41 +1253,69 @@ enum wifi_nrf_status tx_init(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx)
...
+tx_pkt_info_free:
+	for (i = 0; i < fpriv->num_tx_tokens; i++) {
+		wifi_nrf_utils_list_free(fpriv->opriv,
+						fmac_dev_ctx->tx_config.pkt_info_p[i].pkt);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:1300]: Indentation

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c [hunk 8]
```diff
@@ -1311,41 +1253,69 @@ enum wifi_nrf_status tx_init(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx)
...
+	}
+tx_q_setup_free:
+	wifi_nrf_osal_mem_free(fmac_dev_ctx->fpriv->opriv,
+							fmac_dev_ctx->tx_config.pkt_info_p);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:1304]: Indentation

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c [hunk 9]
```diff
@@ -1311,41 +1253,69 @@ enum wifi_nrf_status tx_init(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx)
...
+tx_q_free:
+	for (i = 0; i < WIFI_NRF_FMAC_AC_MAX; i++) {
+		for (j = 0; j < MAX_SW_PEERS; j++) {
+			q_ptr = fmac_dev_ctx->tx_config.data_pending_txq[j][i];
 
-				wifi_nrf_utils_q_free(fpriv->opriv,
-						      q_ptr);
-			}
+			wifi_nrf_utils_q_free(fpriv->opriv,
+							q_ptr);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:1311]: Indentation

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c [hunk 10]
```diff
@@ -1311,41 +1253,69 @@ enum wifi_nrf_status tx_init(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx)
...
-
-		goto out;
 	}
-
-	fmac_dev_ctx->twt_sleep_status = WIFI_NRF_FMAC_TWT_STATE_AWAKE;
-
-	status = WIFI_NRF_STATUS_SUCCESS;
+coal_q_free:
+	wifi_nrf_osal_mem_free(fmac_dev_ctx->fpriv->opriv,
+					fmac_dev_ctx->tx_config.send_pkt_coalesce_count_p);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:1316]: Indentation

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_api.c [hunk 11]
```diff
@@ -1121,18 +1121,19 @@ struct wifi_nrf_hal_dev_ctx *wifi_nrf_hal_dev_add(struct wifi_nrf_hal_priv *hpri
 	wifi_nrf_osal_spinlock_init(hpriv->opriv,
 				    hal_dev_ctx->lock_rx);
 
-	hal_dev_ctx->rx_tasklet = wifi_nrf_osal_tasklet_alloc(hpriv->opriv);
+	hal_dev_ctx->event_tasklet = wifi_nrf_osal_tasklet_alloc(hpriv->opriv,
+		WIFI_NRF_TASKLET_TYPE_BH);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_api.c:1125]: Indentation

File: drivers/wifi/nrf700x/osal/os_if/inc/osal_structs.h [hunk 12]
```diff
@@ -45,6 +45,27 @@ enum wifi_nrf_osal_dma_dir {
...
+/**
+ * enum wifi_nrf_tasklet_type - The type of a tasklet
+ * @WIFI_NRF_TASKLET_TYPE_BH: The tasklet is a bottom half tasklet i.e it is
+ *		scheduled from an interrupt context used for all except
+ *		TX done tasklets.
+ * @WIFI_NRF_TASKLET_TYPE_IRQ: The tasklet is an IRQ tasklet. It is scheduled
+ *		from the Bus ISR, used internally by the SHIM layer.
+ * @WIFI_NRF_TASKLET_TYPE_TX_DONE: The tasklet is a TX done tasklet. It is
+ *		scheduled from the BH tasklet for TX done interrupts.
+ * @WIFI_NRF_TASKLET_TYPE_MAX: The maximum number of tasklet types.
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/os_if/inc/osal_structs.h:57]: Description for WIFI_NRF_TASKLET_TYPE_RX missing

File: drivers/wifi/nrf700x/zephyr/src/zephyr_work.c [hunk 13]
```diff
@@ -77,6 +88,25 @@ static int workqueue_init(void)
 						K_THREAD_STACK_SIZEOF(irq_wq_stack_area),
 						CONFIG_NRF700X_IRQ_WQ_PRIORITY,
 						NULL);
+#ifdef CONFIG_NRF700X_TX_DONE_WQ_ENABLED
+	k_work_queue_init(&zep_wifi_tx_done_q);
+
+	k_work_queue_start(&zep_wifi_tx_done_q,
+						tx_done_wq_stack_area,
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_work.c:95]: Indentation

File: drivers/wifi/nrf700x/zephyr/src/zephyr_work.c [hunk 14]
```diff
@@ -77,6 +88,25 @@ static int workqueue_init(void)
...
+						K_THREAD_STACK_SIZEOF(tx_done_wq_stack_area),
+						CONFIG_NRF700X_TX_DONE_WQ_PRIORITY,
+						NULL);
+#endif /* CONFIG_NRF700X_TX_DONE_WQ_ENABLED */
+
+#ifdef CONFIG_NRF700X_RX_WQ_ENABLED
+	k_work_queue_init(&zep_wifi_rx_q);
+
+	k_work_queue_start(&zep_wifi_rx_q,
+						rx_wq_stack_area,
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_work.c:105]: Indentation

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c (hunk 5 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:1293]: Tried again, aligning to open brace vs. not exceeding 100 lines vs. using tabs as much as possible, its impossible to do this alignment anyother (AFAIK), so, closing this comment as checkpatch is passing.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c (hunk 9 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:1311]: Tried again, aligning to open brace vs. not exceeding 100 lines vs. using tabs as much as possible, its impossible to do this alignment anyother (AFAIK), so, closing this comment as checkpatch is passing.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c (hunk 10 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:1316]: Tried again, aligning to open brace vs. not exceeding 100 lines vs. using tabs as much as possible, its impossible to do this alignment anyother (AFAIK), so, closing this comment as checkpatch is passing.

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_api.c (hunk 11 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_api.c:1125]: Tried again, aligning to open brace vs. not exceeding 100 lines vs. using tabs as much as possible, its impossible to do this alignment anyother (AFAIK), so, closing this comment as checkpatch is passing.

File: drivers/wifi/nrf700x/zephyr/src/zephyr_work.c (hunk 13 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/zephyr/src/zephyr_work.c:95]: Tried again, aligning to open brace vs. not exceeding 100 lines vs. using tabs as much as possible, its impossible to do this alignment anyother (AFAIK), so, closing this comment as checkpatch is passing.

File: drivers/wifi/nrf700x/zephyr/src/zephyr_work.c (hunk 14 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/zephyr/src/zephyr_work.c:105]: Tried again, aligning to open brace vs. not exceeding 100 lines vs. using tabs as much as possible, its impossible to do this alignment anyother (AFAIK), so, closing this comment as checkpatch is passing.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c (hunk 8 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:1304]: Tried again, aligning to open brace vs. not exceeding 100 lines vs. using tabs as much as possible, its impossible to do this alignment anyother (AFAIK), so, closing this comment as checkpatch is passing.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c (hunk 4 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:164]: Tried again, aligning to open brace vs. not exceeding 100 lines vs. using tabs as much as possible, its impossible to do this alignment anyother (AFAIK), so, closing this comment as checkpatch is passing.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/event.c (hunk 3 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/event.c:543]: Tried again, aligning to open brace vs. not exceeding 100 lines vs. using tabs as much as possible, its impossible to do this alignment anyother (AFAIK), so, closing this comment as checkpatch is passing.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/event.c (hunk 2 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/event.c:578]: Tried again, aligning to open brace vs. not exceeding 100 lines vs. using tabs as much as possible, its impossible to do this alignment anyother (AFAIK), so, closing this comment as checkpatch is passing.

**Review by sachinthegreen — APPROVED** [review]

**Review by rado17 — APPROVED** [review]

**Pull request merged**